src/search/task_proxy.cc

```diff
@@ -1,11 +1,60 @@
 #include "task_proxy.h"
 
+#include "axioms.h"
+#include "state_registry.h"
+
 #include "task_utils/causal_graph.h"
 
 #include <iostream>
 
 using namespace std;
 
+State::State(const AbstractTask &task, const StateRegistry &registry,
+             const StateID &id, const PackedStateBin *buffer)
+    : task(&task), registry(&registry), id(id), buffer(buffer), values(nullptr),
+      state_packer(&registry.get_state_packer()),
+      num_variables(registry.get_num_variables()) {
+    assert(id != StateID::no_state);
+    assert(buffer);
+    assert(num_variables == task.get_num_variables());
+}
+
+State::State(const AbstractTask &task, const StateRegistry &registry,
+             const StateID &id, const PackedStateBin *buffer,
+             vector<int> &&values)
+    : State(task, registry, id, buffer) {
+    assert(num_variables == values.size());
+    this->values = make_shared<vector<int>>(move(values));
+}
+
+State::State(const AbstractTask &task, vector<int> &&values)
+    : task(&task), registry(nullptr), id(StateID::no_state), buffer(nullptr),
+      values(make_shared<vector<int>>(move(values))),
+      state_packer(nullptr), num_variables(this->values->size()) {
+    assert(num_variables == task.get_num_variables());
+}
+
+State State::get_unregistered_successor(const OperatorProxy &op) const {
+    assert(!op.is_axiom());
+    //assert(is_applicable(op, state));
+
+    unpack();
```

Comment:
Is it documented that this function implicitly unpacks?

Reply:
If this is the only such place, perhaps this should be an assertion instead (and we should not unpack). But then we need to document that the state must be unpacked.